File: kotlin-ktor-sample/src/main/kotlin/com/example/repositories/ListingRepository.kt
package com.example.repositories

import com.example.models.Listing
import com.example.models.Price
import com.example.models.Location
import com.example.models.SearchFilter
import com.example.models.FilterOperator
import java.math.BigDecimal
import java.sql.*
import java.util.UUID
import javax.sql.DataSource

class ListingRepository(private val dataSource: DataSource) : IListingRepository {

    override fun getAll(page: Int, pageSize: Int): List<Listing> {
        val sql = """
            SELECT 
                listing_id,
                name,
                description,
                price_currency,
                price_amount,
                category,
                location_country,
                location_municipality,
                location_geohash
            FROM listings 
            ORDER BY created_at DESC
            LIMIT ? OFFSET ?
        """.trimIndent()

        return dataSource.connection.use { connection ->
            connection.prepareStatement(sql).use { statement ->
                val offset = (page - 1) * pageSize
                statement.setInt(1, pageSize)
                statement.setInt(2, offset)
                
                statement.executeQuery().use { resultSet ->
                    val listings = mutableListOf<Listing>()
                    while (resultSet.next()) {
                        listings.add(mapToListing(resultSet))
                    }
                    listings
                }
            }
        }
    }

    override fun getTotalCount(): Int {
        val sql = "SELECT COUNT(*) FROM listings"
        
        return dataSource.connection.use { connection ->
            connection.prepareStatement(sql).use { statement ->
                statement.executeQuery().use { resultSet ->
                    resultSet.next()
                    resultSet.getInt(1)
                }
            }
        }
    }

    override fun getById(id: UUID): Listing? {
        val sql = """
            SELECT 
                listing_id,
                name,
                description,
                price_currency,
                price_amount,
                category,
                location_country,
                location_municipality,
                location_geohash
            FROM listings 
            WHERE listing_id = ?
        """.trimIndent()

        return dataSource.connection.use { connection ->
            connection.prepareStatement(sql).use { statement ->
                statement.setObject(1, id)
                
                statement.executeQuery().use { resultSet ->
                    if (resultSet.next()) {
                        mapToListing(resultSet)
                    } else {
                        null
                    }
                }
            }
        }
    }

    override fun create(listing: Listing): Listing {
        val sql = """
            INSERT INTO listings (
                listing_id, name, description, price_currency, price_amount, 
                category, location_country, location_municipality, location_geohash
            ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
        """.trimIndent()

        val newListing = listing.copy(listingId = UUID.randomUUID())

        dataSource.connection.use { connection ->
            connection.autoCommit = false
            try {
                connection.prepareStatement(sql).use { statement ->
                    statement.setObject(1, newListing.listingId)
                    statement.setString(2, newListing.name)
                    statement.setString(3, newListing.description)
                    statement.setString(4, newListing.price.currency)
                    statement.setBigDecimal(5, newListing.price.amount)
                    statement.setString(6, newListing.category)
                    statement.setString(7, newListing.location.country)
                    statement.setString(8, newListing.location.municipality)
                    statement.setString(9, newListing.location.geohash)
                    
                    statement.executeUpdate()
                }
                connection.commit()
            } catch (e: Exception) {
                connection.rollback()
                throw e
            }
        }

        return newListing
    }

    override fun update(listing: Listing): Listing? {
        val sql = """
            UPDATE listings SET 
                name = ?,
                description = ?,
                price_currency = ?,
                price_amount = ?,
                category = ?,
                location_country = ?,
                location_municipality = ?,
                location_geohash = ?,
                updated_at = CURRENT_TIMESTAMP
            WHERE listing_id = ?
        """.trimIndent()

        return dataSource.connection.use { connection ->
            connection.autoCommit = false
            try {
                connection.prepareStatement(sql).use { statement ->
                    statement.setString(1, listing.name)
                    statement.setString(2, listing.description)
                    statement.setString(3, listing.price.currency)
                    statement.setBigDecimal(4, listing.price.amount)
                    statement.setString(5, listing.category)
                    statement.setString(6, listing.location.country)
                    statement.setString(7, listing.location.municipality)
                    statement.setString(8, listing.location.geohash)
                    statement.setObject(9, listing.listingId)
                    
                    val rowsAffected = statement.executeUpdate()
                    connection.commit()
                    if (rowsAffected > 0) listing else null
                }
            } catch (e: Exception) {
                connection.rollback()
                throw e
            }
        }
    }

    override fun delete(id: UUID): Boolean {
        val sql = "DELETE FROM listings WHERE listing_id = ?"
        
        return dataSource.connection.use { connection ->
            connection.autoCommit = false
            try {
                connection.prepareStatement(sql).use { statement ->
                    statement.setObject(1, id)
                    val rowsAffected = statement.executeUpdate()
                    connection.commit()
                    rowsAffected > 0
                }
            } catch (e: Exception) {
                connection.rollback()
                throw e
            }
        }
    }

    override fun exists(id: UUID): Boolean {
        val sql = "SELECT COUNT(*) FROM listings WHERE listing_id = ?"
        
        return dataSource.connection.use { connection ->
            connection.prepareStatement(sql).use { statement ->
                statement.setObject(1, id)
                
                statement.executeQuery().use { resultSet ->
                    resultSet.next()
                    resultSet.getInt(1) > 0
                }
            }
        }
    }

    override fun searchListings(filters: List<SearchFilter>, page: Int, pageSize: Int): List<Listing> {
        val baseQuery = """
            SELECT 
                listing_id,
                name,
                description,
                price_currency,
                price_amount,
                category,
                location_country,
                location_municipality,
                location_geohash
            FROM listings
        """.trimIndent()

        val (whereClause, parameters) = buildWhereClause(filters)
        val sql = if (whereClause.isNotEmpty()) {
            "$baseQuery WHERE $whereClause ORDER BY created_at DESC LIMIT ? OFFSET ?"
        } else {
            "$baseQuery ORDER BY created_at DESC LIMIT ? OFFSET ?"
        }

        return dataSource.connection.use { connection ->
            connection.prepareStatement(sql).use { statement ->
                var paramIndex = 1
                
                // Set filter parameters
                for ((_, value) in parameters) {
                    statement.setObject(paramIndex++, value)
                }
                
                // Set pagination parameters
                val offset = (page - 1) * pageSize
                statement.setInt(paramIndex++, pageSize)
                statement.setInt(paramIndex, offset)
                
                statement.executeQuery().use { resultSet ->
                    val listings = mutableListOf<Listing>()
                    while (resultSet.next()) {
                        listings.add(mapToListing(resultSet))
                    }
                    listings
                }
            }
        }
    }

    override fun getSearchResultCount(filters: List<SearchFilter>): Int {
        val baseQuery = "SELECT COUNT(*) FROM listings"
        val (whereClause, parameters) = buildWhereClause(filters)
        val sql = if (whereClause.isNotEmpty()) {
            "$baseQuery WHERE $whereClause"
        } else {
            baseQuery
        }

        return dataSource.connection.use { connection ->
            connection.prepareStatement(sql).use { statement ->
                var paramIndex = 1
                
                // Set filter parameters
                for ((_, value) in parameters) {
                    statement.setObject(paramIndex++, value)
                }
                
                statement.executeQuery().use { resultSet ->
                    resultSet.next()
                    resultSet.getInt(1)
                }
            }
        }
    }

    private fun buildWhereClause(filters: List<SearchFilter>): Pair<String, List<Pair<String, Any>>> {
        if (filters.isEmpty()) {
            return Pair("", emptyList())
        }

        val conditions = mutableListOf<String>()
        val parameters = mutableListOf<Pair<String, Any>>()

        for (filter in filters) {
            val columnName = mapFilterFieldToColumn(filter.field)
            
            when (filter.operator) {
                FilterOperator.CONTAINS -> {
                    conditions.add("LOWER($columnName) LIKE LOWER(?)")
                    parameters.add(Pair(columnName, "%${filter.value}%"))
                }
                FilterOperator.EQUALS -> {
                    conditions.add("$columnName = ?")
                    parameters.add(Pair(columnName, filter.value))
                }
            }
        }

        return Pair(conditions.joinToString(" AND "), parameters)
    }

    private fun mapFilterFieldToColumn(field: String): String {
        return when (field) {
            "name" -> "name"
            "description" -> "description"
            "category" -> "category"
            "location.country" -> "location_country"
            "location.municipality" -> "location_municipality"
            else -> throw IllegalArgumentException("Unsupported filter field: $field")
        }
    }

    private fun mapToListing(resultSet: ResultSet): Listing {
        return Listing(
            listingId = resultSet.getObject("listing_id", UUID::class.java),
            name = resultSet.getString("name"),
            description = resultSet.getString("description"),
            price = Price(
                currency = resultSet.getString("price_currency"),
                amount = resultSet.getBigDecimal("price_amount")
            ),
            category = resultSet.getString("category"),
            location = Location(
                country = resultSet.getString("location_country"),
                municipality = resultSet.getString("location_municipality"),
                geohash = resultSet.getString("location_geohash")
            )
        )
    }
}

File: csharp-aspnetcore-sample/Repositories/ListingRepository.cs
using System.Data;
using csharp_aspnetcore_sample.Models;
using Dapper;
using Npgsql;

namespace csharp_aspnetcore_sample.Repositories;

public class ListingRepository : IListingRepository
{
    private readonly string _connectionString;

    public ListingRepository(IConfiguration configuration)
    {
        _connectionString = configuration.GetConnectionString("DefaultConnection") 
            ?? throw new InvalidOperationException("Connection string 'DefaultConnection' not found.");
    }

    private IDbConnection CreateConnection() => new NpgsqlConnection(_connectionString);

    public async Task<IEnumerable<Listing>> GetAllAsync(int page, int pageSize)
    {
        const string sql = @"
            SELECT 
                listing_id as ListingId,
                name as Name,
                description as Description,
                price_currency as Currency,
                price_amount as Amount,
                category as Category,
                location_country as Country,
                location_municipality as Municipality,
                location_geohash as Geohash
            FROM listings 
            ORDER BY created_at DESC
            LIMIT @PageSize OFFSET @Offset";

        using var connection = CreateConnection();
        var offset = (page - 1) * pageSize;
        
        var result = await connection.QueryAsync(sql, new { PageSize = pageSize, Offset = offset });
        
        return result.Select(MapToListing);
    }

    public async Task<int> GetTotalCountAsync()
    {
        const string sql = "SELECT COUNT(*) FROM listings";
        
        using var connection = CreateConnection();
        return await connection.QuerySingleAsync<int>(sql);
    }

    public async Task<Listing?> GetByIdAsync(Guid id)
    {
        const string sql = @"
            SELECT 
                listing_id as ListingId,
                name as Name,
                description as Description,
                price_currency as Currency,
                price_amount as Amount,
                category as Category,
                location_country as Country,
                location_municipality as Municipality,
                location_geohash as Geohash
            FROM listings 
            WHERE listing_id = @Id";

        using var connection = CreateConnection();
        var result = await connection.QuerySingleOrDefaultAsync(sql, new { Id = id });
        
        return result != null ? MapToListing(result) : null;
    }

    public async Task<Listing> CreateAsync(Listing listing)
    {
        const string sql = @"
            INSERT INTO listings (
                listing_id, name, description, price_currency, price_amount, 
                category, location_country, location_municipality, location_geohash
            ) VALUES (
                @ListingId, @Name, @Description, @Currency, @Amount, 
                @Category, @Country, @Municipality, @Geohash
            )";

        listing.ListingId = Guid.NewGuid();

        using var connection = CreateConnection();
        await connection.ExecuteAsync(sql, new
        {
            ListingId = listing.ListingId,
            Name = listing.Name,
            Description = listing.Description,
            Currency = listing.Price.Currency,
            Amount = listing.Price.Amount,
            Category = listing.Category.ToString(),
            Country = listing.Location.Country,
            Municipality = listing.Location.Municipality,
            Geohash = listing.Location.Geohash
        });

        return listing;
    }

    public async Task<Listing?> UpdateAsync(Listing listing)
    {
        const string sql = @"
            UPDATE listings SET 
                name = @Name,
                description = @Description,
                price_currency = @Currency,
                price_amount = @Amount,
                category = @Category,
                location_country = @Country,
                location_municipality = @Municipality,
                location_geohash = @Geohash,
                updated_at = CURRENT_TIMESTAMP
            WHERE listing_id = @ListingId";

        using var connection = CreateConnection();
        var rowsAffected = await connection.ExecuteAsync(sql, new
        {
            ListingId = listing.ListingId,
            Name = listing.Name,
            Description = listing.Description,
            Currency = listing.Price.Currency,
            Amount = listing.Price.Amount,
            Category = listing.Category.ToString(),
            Country = listing.Location.Country,
            Municipality = listing.Location.Municipality,
            Geohash = listing.Location.Geohash
        });

        return rowsAffected > 0 ? listing : null;
    }

    public async Task<bool> DeleteAsync(Guid id)
    {
        const string sql = "DELETE FROM listings WHERE listing_id = @Id";
        
        using var connection = CreateConnection();
        var rowsAffected = await connection.ExecuteAsync(sql, new { Id = id });
        
        return rowsAffected > 0;
    }

    public async Task<bool> ExistsAsync(Guid id)
    {
        const string sql = "SELECT COUNT(*) FROM listings WHERE listing_id = @Id";
        
        using var connection = CreateConnection();
        var count = await connection.QuerySingleAsync<int>(sql, new { Id = id });
        
        return count > 0;
    }

    public async Task<(IEnumerable<Listing> Items, int TotalCount)> SearchAsync(List<Filter> filters, int page, int pageSize)
    {
        var baseSql = @"
            SELECT 
                listing_id as ListingId,
                name as Name,
                description as Description,
                price_currency as Currency,
                price_amount as Amount,
                category as Category,
                location_country as Country,
                location_municipality as Municipality,
                location_geohash as Geohash
            FROM listings";

        var countSql = "SELECT COUNT(*) FROM listings";
        
        var whereConditions = new List<string>();
        var parameters = new DynamicParameters();
        
        BuildWhereClause(filters, whereConditions, parameters);
        
        if (whereConditions.Count > 0)
        {
            var whereClause = " WHERE " + string.Join(" AND ", whereConditions);
            baseSql += whereClause;
            countSql += whereClause;
        }
        
        baseSql += " ORDER BY created_at DESC";
        
        var offset = (page - 1) * pageSize;
        parameters.Add("@PageSize", pageSize);
        parameters.Add("@Offset", offset);
        
        var paginatedSql = baseSql + " LIMIT @PageSize OFFSET @Offset";
        
        using var connection = CreateConnection();
        
        var itemsTask = await connection.QueryAsync(paginatedSql, parameters);
        var totalCount = await connection.QuerySingleAsync<int>(countSql, parameters);
        
        var items = itemsTask.Select(MapToListing);
        
        return (items, totalCount);
    }

    private static void BuildWhereClause(List<Filter> filters, List<string> whereConditions, DynamicParameters parameters)
    {
        for (int i = 0; i < filters.Count; i++)
        {
            var filter = filters[i];
            var paramName = $"param{i}";
            
            switch (filter.Field.ToLowerInvariant())
            {
                case "name":
                    if (filter.Operator.ToLowerInvariant() == "contains")
                    {
                        whereConditions.Add($"LOWER(name) LIKE LOWER(@{paramName})");
                        parameters.Add($"@{paramName}", $"%{filter.Value}%");
                    }
                    break;
                    
                case "description":
                    if (filter.Operator.ToLowerInvariant() == "contains")
                    {
                        whereConditions.Add($"LOWER(description) LIKE LOWER(@{paramName})");
                        parameters.Add($"@{paramName}", $"%{filter.Value}%");
                    }
                    break;
                    
                case "category":
                    if (filter.Operator.ToLowerInvariant() == "equals")
                    {
                        whereConditions.Add($"category = @{paramName}");
                        parameters.Add($"@{paramName}", filter.Value.ToString());
                    }
                    break;
                    
                case "location.country":
                    if (filter.Operator.ToLowerInvariant() == "contains")
                    {
                        whereConditions.Add($"LOWER(location_country) LIKE LOWER(@{paramName})");
                        parameters.Add($"@{paramName}", $"%{filter.Value}%");
                    }
                    break;
                    
                case "location.municipality":
                    if (filter.Operator.ToLowerInvariant() == "contains")
                    {
                        whereConditions.Add($"LOWER(location_municipality) LIKE LOWER(@{paramName})");
                        parameters.Add($"@{paramName}", $"%{filter.Value}%");
                    }
                    break;
            }
        }
    }

    private static Listing MapToListing(dynamic row)
    {
        var categoryString = (string)row.category;
        var category = Enum.Parse<Category>(categoryString.Replace(" & ", "And").Replace(" ", "").Replace(",", ""));

        return new Listing
        {
            ListingId =(Guid)row.listingid,
            Name = (string)row.name,
            Description = (string)row.description,
            Price = new Price
            {
                Currency = (string)row.currency,
                Amount = (decimal)row.amount
            },
            Category = category,
            Location = new Location
            {
                Country = (string)row.country,
                Municipality = (string)row.municipality,
                Geohash = (string)row.geohash
            }
        };
    }
}